### PR TITLE
fix: notch works on stadium

### PR DESCRIPTION
**Pull request by jarrodcolburn**
https://github.com/flutter/flutter/issues/21650

### CODE REVIEW

**Review by goderbauer — COMMENTED** [review]
Thanks for your contribution!

In order for us to accept this PR into our repository you'll have to add tests to ensure that this functionality keeps working as expected in the future.

File: packages/flutter/lib/src/painting/notched_shapes.dart [hunk 1]
```diff
@@ -88,22 +88,40 @@ class CircularNotchedRectangle implements NotchedShape {
...
+    if (guest.height == guest.width) {
+      // circle
+      path.arcToPoint(
         p[3],
         radius: Radius.circular(notchRadius),
         clockwise: false,
-      )
+      );
+    } else {
+      // stadium
```
[the remark this answers is not in the document]
Comment (goderbauer) [reply packages/flutter/lib/src/painting/notched_shapes.dart]: Looking at other parts of our code, we already have a StadiumBorder [1] in the code base. Based on that, this code be the StadiumNotchedRectangle.

[1] https://master-docs.flutter.io/flutter/painting/StadiumBorder-class.html

[the remark this answers is not in the document]
Comment (JarrodCColburn) [reply packages/flutter/lib/src/painting/notched_shapes.dart]: True, I made a very similar comment myself https://github.com/flutter/flutter/issues/21650#issuecomment-457491753 . But one thing that I think might possibly make this different is that the `ShapeBorder` is describing a single object, whereas these notches are describing an intersection between two objects. So far `NotchedShape` looks to only have the single implementer (`CircularNotchedRectangle`). This weekend I'd like to implement another ~3 `NotchedShape` shapes. Why I don't I do that, then we could look at them next to each other for their similarities/differences which may make us come to a naming convention that will be good forward going?

[the remark this answers is not in the document]
Comment (goderbauer) [reply packages/flutter/lib/src/painting/notched_shapes.dart]: Sounds good.

**Comment by Hixie** [comment]
I think https://github.com/flutter/flutter/pull/27487 is probably a better way to solve this.

**Comment by goderbauer** [comment]
@JarrodCColburn https://github.com/flutter/flutter/pull/27487 got merged and is available on Master. Does this solve your issue?

**Comment by goderbauer** [comment]
I am going to close this one since #27487 landed and should be a better fix for this. Please let us know if this doesn't work for you.

**Comment by lucaslcode** [comment]
@goderbauer This isn't fixed still, `AutomaticNotchedShape` just exactly subtracts one shape from another, doesn't give a nice curved radius with padding like `CircularNotchedRectangle` and also doesn't work with text directionality so isn't very easy to use.

**Pull request closed without merging**